neqo-qpack/src/decoder_instructions.rs

```diff
@@ -0,0 +1,231 @@
+use crate::{Error, Res};
+use neqo_common::{qdebug, qtrace};
+use std::mem;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+pub enum DecoderInstruction {
+    InsertCountIncrement { increment: u64 },
+    HeaderAck { stream_id: u64 },
+    StreamCancellation { stream_id: u64 },
+    NoInstruction,
+}
+
+impl DecoderInstruction {
+    fn get_instruction(b: u8) -> Self {
+        if DECODER_HEADER_ACK.cmp_prefix(b) {
+            Self::HeaderAck { stream_id: 0 }
+        } else if DECODER_STREAM_CANCELLATION.cmp_prefix(b) {
+            Self::StreamCancellation { stream_id: 0 }
+        } else if DECODER_INSERT_COUNT_INCREMENT.cmp_prefix(b) {
+            Self::InsertCountIncrement { increment: 0 }
+        } else {
+            unreachable!("The above patterns match everything.");
+        }
+    }
+
+    pub(crate) fn marshal(&self, enc: &mut QPData) {
+        match self {
+            Self::InsertCountIncrement { increment } => {
+                enc.encode_prefixed_encoded_int(DECODER_INSERT_COUNT_INCREMENT, *increment);
+            }
+            Self::HeaderAck { stream_id } => {
+                enc.encode_prefixed_encoded_int(DECODER_HEADER_ACK, *stream_id);
+            }
+            Self::StreamCancellation { stream_id } => {
+                enc.encode_prefixed_encoded_int(DECODER_STREAM_CANCELLATION, *stream_id);
+            }
+            Self::NoInstruction => {}
+        }
+    }
+}
+
+#[derive(Debug)]
+enum DecoderInstructionReaderState {
+    ReadInstruction,
+    ReadInt { reader: IntReader },
+}
+
+#[derive(Debug)]
+pub struct DecoderInstructionReader {
+    state: DecoderInstructionReaderState,
+    instruction: DecoderInstruction,
+}
+
+impl ::std::fmt::Display for DecoderInstructionReader {
+    fn fmt(&self, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+        write!(f, "InstructionReader")
+    }
+}
+
+impl DecoderInstructionReader {
+    pub fn new() -> Self {
+        Self {
+            state: DecoderInstructionReaderState::ReadInstruction,
+            instruction: DecoderInstruction::NoInstruction,
+        }
+    }
+
+    pub fn read_instructions<R: ReadByte>(
+        &mut self,
+        recv: &mut R,
+    ) -> Res<Option<DecoderInstruction>> {
+        qdebug!([self], "read a new instraction");
+        loop {
+            match &mut self.state {
+                DecoderInstructionReaderState::ReadInstruction => match recv.read_byte() {
+                    Ok(b) => {
+                        self.instruction = DecoderInstruction::get_instruction(b);
+                        self.state = DecoderInstructionReaderState::ReadInt {
+                            reader: IntReader::new(
+                                b,
+                                if DECODER_HEADER_ACK.cmp_prefix(b) {
+                                    DECODER_HEADER_ACK.len()
+                                } else {
+                                    DECODER_STREAM_CANCELLATION.len()
+                                },
```

Comment:
It seems like this might be an opportunity to put a method on Prefix that handles this.  Maybe `DECODER_HEADER_ACK.int_reader(b).or_else(|| DECODER_STREAM_CANCELLATION.int_reader(b)).unwrap()`

Or you could do that on IntReader: `IntReader::make(&[DECODER_HEADER_ACK, DECODER_STREAM_CANCELLATION])` where the list is iterated over until the prefix matches.